fetch_index.py

```diff
@@ -0,0 +1,421 @@
+import time
+from sharepoint_utils import print_and_log
+from urllib.parse import urljoin
+from elastic_enterprise_search import WorkplaceSearch
+from checkpointing import Checkpoint
+from sharepoint_client import SharePoint
+from configuration import Configuration
+import logger_manager as log
+from usergroup_permissions import Permissions
+from datetime import datetime
+
+logger = log.setup_logging("workplace_search_index")
+SITE = "site"
+LIST = "list"
+ITEM = "item"
+SITES = "sites"
+LISTS = "lists"
+ITEMS = "items"
+
+
+def check_response(response, error_message, exception_message, param_name):
+    if not response:
+        logger.error(error_message)
+        return None if param_name == SITES else []
+    try:
+        response_data = response.json()
+        response_data = response_data.get("d", {}).get("results")
+        return response_data
+    except ValueError as exception:
+        logger.exception("%s Error: %s" % (exception_message, exception))
+        return None if param_name == SITES else []
+
+
+class FetchIndex:
+
+    def __init__(self, data, query):
+        logger.info("Initializing the Indexing class")
+        self.is_error = False
+        self.ws_host = data.get("enterprise_search.host_url")
+        self.ws_token = data.get("workplace_search.access_token")
+        self.ws_source = data.get("workplace_search.source_id")
+        self.sharepoint_host = data.get("sharepoint.host_url")
+        self.objects = data.get("objects")
+        self.site_collections = data.get("sharepoint.site_collections")
+        self.query = query
+        self.start_time = data.get("start_time")
+        self.end_time = data.get("end_time")
+        self.checkpoint = Checkpoint(logger, data)
+        self.sharepoint_client = SharePoint(logger)
+        self.permissions = Permissions(logger, self.sharepoint_client)
+        self.ws_client = WorkplaceSearch(self.ws_host, http_auth=self.ws_token)
+
+    def index_sites(self, collection, index):
+        rel_url = urljoin(
+            self.sharepoint_host, f"/sites/{collection}/_api/web/webs"
+        )
+        logger.info("Fetching the sites detail from url: %s" % (rel_url))
+        response = self.sharepoint_client.get(rel_url, self.query)
+
+        response_data = check_response(
+            response,
+            "Could not fetch the sites, url: %s" % (rel_url),
+            "Error while parsing the get sites response from url: %s."
+            % (rel_url),
+            SITES,
+        )
+
+        logger.info(
+            "Successfuly fetched and parsed the sites response from SharePoint"
+        )
+        logger.info("Indexing the sites to the Workplace")
+        schema = ['Created', 'Id', 'LastItemModifiedDate',
```

Comment:
because we do `field.lower()`, the document that gets indexed in Workplace Search will have a structure similar to:
```
{
  "created": ...,
  "id": ...,
  "lastitemmodifieddate': ...,
  "serverrelativeurl": ...,
  "title": ...,
  "url"...
}
```
we would want these to match how documents are indexed in other Workplace Search connectors, so:
`Created` -> `created_at`
`id` stays the same
`LastItemModifiedDate` -> `last_updated`
`ServerRelativeUrl` -> `server_relative_url`
`title` stays the same
`url` stays the same

for example these [docs](https://www.elastic.co/guide/en/workplace-search/current/workplace-search-sharepoint-online-connector.html#sharepoint-fields) include a list of the current fields we have for SharePoint Online.

I would also strongly suggest that for each type of indexed objects we do this conversion in separate methods, that could be defined in a separate file e.g. adapter.py.
This makes it easier to understand and modify what fields are indexed if someone wants to customise the behaviour.

Reply:
Done